src/core/Reactor.cpp

```diff
@@ -78,68 +82,92 @@ namespace core {
 		if (server.listen() == false)
 			throw std::runtime_error("A VirtualServer couldn't acquire its socket!");
 
-		m_virtual_servers.push_back(server);
+		m_vServers.push_back(server);
+	}
+
+	bool Reactor::addVirtualServers(config::t_config_data &configData) {
+		for (size_t i = 0; i < configData.servers.size(); ++i) {
+			http::VirtualServer server(configData.servers.at(i));
+			if (server.getSocket().init() == false) {
+				m_vServers.clear();
+				std::cout << "Failed to initialize socket for server "
+						  << configData.servers.at(i).server_names.at(0) << std::endl;
+				return false;
+			}
+			for (size_t j = 0; j < configData.servers.at(i).locations.size(); ++j) {
+				// server.addLocation(configData.servers.at(i).locations.at(j));
+			}
+			m_vServers.push_back(server);
+		}
+		return true;
 	}
 
 	bool Reactor::removeVirtualServer(
 		std::vector<http::VirtualServer>::iterator it) {
-		m_virtual_servers.erase(it);
+		m_vServers.erase(it);
 		return true;
 	}
 
-	void Reactor::registerHandler(int fd, core::RequestHandler *handler, uint32_t events) throw(std::runtime_error) {
+	void Reactor::registerHandler(int fd, AHandler *handler, const HandlerContext &ctx, uint32_t events) {
 		t_event event;
 
-		event.data.fd = fd;
 		event.events = events;
+		event.data.ptr = new EventData(handler, ctx); // note: decided to store data int the event ptr to avoid later lookup
 		if (epoll_ctl(m_epoll_master_fd, EPOLL_CTL_ADD, fd, &event) < 0) {
 			throw std::runtime_error("Crud! Couldn't register event handler!");
 		}
-		m_event_handlers[fd] = handler;
 	}
 
 	void Reactor::unregisterHandler(int fd) throw(std::runtime_error) {
-		if (epoll_ctl(m_epoll_master_fd, EPOLL_CTL_DEL, fd, NULL) < 0) {
+		t_event event;
+
+		if (epoll_ctl(m_epoll_master_fd, EPOLL_CTL_DEL, fd, &event) < 0) {
 			throw std::runtime_error("Пиздец! Couldn't unregister event handler!");
 		}
-		m_event_handlers.erase(fd);
+
+		EventData *data = reinterpret_cast<EventData *>(event.data.ptr);
+		delete data->handler;
+		delete data;
 	}
 
 	void Reactor::react() {
 		t_event events[MAX_EVENTS];
 
 		while (true) {
+			acceptNewConnections();
+
 			int nEvents = epoll_wait(m_epoll_master_fd, events, MAX_EVENTS, 60);
 			if (nEvents < 0) {
-				throw std::runtime_error("Epoll dun goofed up!");
+				throw std::runtime_error("epoll_wait failed");
```

Comment:
Testing this, I seem to be getting a new connection every time I reload the browser tab. Though I couldn't verify whether it actually *closes* the connection - it has to, I think, but I couldn't get it to print a confirmation when doing this. Mostly just a weird thing I found notable, it's probably fine.

Reply:
I am not really sure where to close the connection. Should I always close the connection after sending the response. When to close this socket ahhh. Currently closing when IOHandler is completed

Reply:
Pretty sure this depends on what's going through the connection. Generally speaking, the server only closes the connection on error or timeout, though otherwise afaik the connection usually stays open until the client closes it (or times out).